### ISSUE TITLE
Add bidi support

### ISSUE THREAD

**Issue by ahangarha** [markedjs/marked#2304]
**What pain point are you perceiving?.**
As we all know, Markdown is supposed to be minimal. We get into issue when we want to make some RTL document with Markdown. It become more complicated when we get some document containing both RTL and LTR text which means imposing one general direction to whole rendered document won't help.

There are some solutions:

- Add some specific tag for it like `{rtl}` - not good because this is against the very idea of Markdown
- Make some configuration out of the document - not good since it won't work with mixed content
- Hack the parser by adding html tags like `<dir dir="rtl">` to that particular section - bad idea since it won't be Markdown any more
- Add bidirectional text (bidi) support

**Describe the solution you'd like**
I am for the last solution: **bidi support**. It doesn't need any alteration or injection in md file at all. One can still write the document in the regular way (of course the editor should support bidi to show the text in correct direction. gedit and kate support it very well) What we need to do is to do little modification in the parser/renderer.

Basically we need to add `dir="auto"` to all elements containing text, meaning all paragraphs, headings, blockquote, code/pre. If in any place we are giving padding or margin, we should use `start|end` instead of `left|right|. and the tricky part is for lists. we should add `dir="auto"` to `ul|ol` as well.

Gitlab has done this for more than two years. Github added this just few weeks ago. And I think this is what all Markdown parsers should do.

As an example look at the text below. Both have same words but the first one starts with latin and the second one starts with Persian letters. As the result, you see `left` and `راست` on correct side.

-----------
left راست

راست left

-----------

If this feature get implemented, it will solve #694 as well.

**Comment by UziTech** [comment]
Is there a spec sheet on this so we can make sure it acts the same as github?

**Comment by ahangarha** [comment]
> Is there a spec sheet on this so we can make sure it acts the same as github?

I don't think there be any spec sheet. But If you want, I can work on it.

I think the following md file is a good reference. It has almost everything would exist in a typical md document. And it is getting rendered perfectly both on Gitlab and now on Github:

- [on Gitlab](https://framagit.org/ahangarha/modern-js-cheatsheet/-/blob/Persian-translation/translations/fa-IR.md)
- [on Github](https://github.com/ahangarha/modern-js-cheatsheet/blob/Persian-translation/translations/fa-IR.md)

**Comment by UziTech** [comment]
If you want to create a PR that would be very helpful.

The difficult thing about this is that the GFM spec doesn't say anything about `dir="auto"` so the tests generated by those specs will fail. I think we will have to add an option but eventually we want to move the options to extensions so maybe this should be an extension from the start?

**Comment by UziTech** [comment]
I started creating an extension at https://github.com/markedjs/marked-bidi.

I haven't released it yet but I think it is working how it is supposed to.

**Comment by ahangarha** [comment]
> it looks like github doesn't add `dir="auto"` to code or tables.

They have some issue with code since they put each line in a dedicated span which needs some css styling to be rendered properly. Also for table there is some issue with `display: table` for which I couldn't find any workaround than adding `dir="auto"` to the able and then adding `margin: auto` to put the teble at the center.

Rest are fine.

I would suggest one thing: Let's work with the new package for sometime and if we found it works well then merge it with the main repo and add an option to enable/disable bidi support.

**Comment by UziTech** [comment]
> then merge it with the main repo and add an option to enable/disable bidi support.

We are trying to get away from having too many options and instead make extensions able to do anything they need.

I'm going to close this as https://www.npmjs.com/package/marked-bidi seems to work pretty well.
